src/Microsoft.Health.Fhir.Liquid.Converter/Processors/JsonToHl7v2Processor.cs

```diff
@@ -60,31 +61,89 @@ public Hl7v2Data GenerateHL7Message(JObject transformations)
         {
             var hl7Segments = new List<Hl7v2Segment>();
 
-            // validate that the template has a messageDefinition section
+            // Validate that the template has a messageDefinition section
             if (transformations["messageDefinition"] == null)
             {
                 throw new RenderException(
-                    FhirConverterErrorCode.TemplateRenderingError,
-                    "The messageDefinition section was not found in the conversion template");
+                    FhirConverterErrorCode.PropertyNotFound,
+                    "The messageDefinition JSON property was not found in the conversion template");
             }
 
-            // create HL7v2 segments
+            // Create HL7v2 segments based on the liquid template the user supplied
             for (int i = 0; i < transformations["messageDefinition"].Count(); i++)
             {
-                var dataResolution = transformations["messageDefinition"][i] as JObject;
-                var segmentName = dataResolution.Properties().First().Name;
+                var dataResolutionForSegment = transformations["messageDefinition"][i] as JObject;
 
-                var fields = new List<Hl7v2Field>();
+                if (dataResolutionForSegment == null)
+                {
+                    throw new RenderException(
+                        FhirConverterErrorCode.TemplateSyntaxError,
+                        $"The segment mapping in position {i} specified in the messageDefinition is not a valid object");
+                }
+
+                var segmentName = string.Empty;
+                try
+                {
+                    segmentName = dataResolutionForSegment.Properties().First().Name;
+                    segmentName = segmentName.ToUpper();
+                }
+                catch
+                {
+                    throw new RenderException(
+                        FhirConverterErrorCode.TemplateSyntaxError,
+                        $"The segment name in position {i} specified in the messageDefinition does not exist or is invalid");
+                }
```

Comment:
nit: Generally using if/else to evaluate null is more efficient than generating exceptions. It sounds like in this case, we are expecting null ref as the only exception, so I think checking for that instead (using ?. to safely access in case of null) as you have above is better.